src/minirest_body.erl

```diff
@@ -0,0 +1,245 @@
+            %% remove \r\n for first part
+            {ok, Headers, RealBodyBinary}
+    end;
+form_data_encoder({form_data, Data}) ->
+    Response = {
+        ?RESPONSE_CODE_INTERNAL_SERVER_ERROR,
+        #{<<"content-type">> => <<"text/plain">>},
+        list_to_binary(io_lib:format("invalid form-data term: ~p", [Data]))
+    },
+    {response, Response}.
+
+form_data_loop_encoder([], Boundary, Res) ->
+    Headers = #{<<"content-type">> => <<"multipart/form-data, boundary=", Boundary/binary>>},
+    {ok, Headers, iolist_to_binary(lists:reverse(Res))};
+
+form_data_loop_encoder([Part | Tail], Boundary, Res) ->
+    case form_data_part(Part, Boundary) of
+        {response, Response} ->
+            {response, Response};
+        PartBinary ->
+            form_data_loop_encoder(Tail, Boundary, [PartBinary | Res])
+    end.
+
+%% form-data file
+form_data_part({file, Path}, Boundary) ->
+    case filelib:is_file(Path) andalso file:read_file_info(Path) of
+        {ok, _} ->
+            case file:read_file(Path) of
+                {ok, Binary} ->
+                    FileName = filename:basename(Path),
+                    form_data_part({file_binary, FileName, Binary}, Boundary);
+                {error, Reason} ->
+                    error_form_data_response({read_file, {Path, Reason}})
+            end;
+        {error, Reason} ->
+            error_form_data_response({bad_file, {Path, Reason}});
+        false ->
+            error_form_data_response({bad_file, Path})
+    end;
+form_data_part({file_binary, FileName, FileBinary}, Boundary) ->
+    NameBinary = to_binary(FileName),
+    FileTypeHeaders = file_content_type_headers(FileName),
+    Headers0 = #{<<"Content-Disposition">> =>
+        <<"form-data; name=\"", NameBinary/binary, "\"; filename=\"", NameBinary/binary, "\"">>},
+    Headers = maps:merge(Headers0, FileTypeHeaders),
+    PartHeaders = cow_multipart:part(Boundary, Headers),
+    iolist_to_binary([PartHeaders, FileBinary]);
+
+%% key & value
+form_data_part({Key, Value}, Boundary) ->
+    case {to_binary(Key), to_binary(Value)} of
+        {{error, {not_support, Key}}, {error, {not_support, Value}}} ->
+            error_form_data_response({Key, Value});
+        {{error, {not_support, Key}}, _} ->
+            error_form_data_response({bad_key, Key});
+        {_, {error, {not_support, Value}}} ->
+            error_form_data_response({bad_value, Value});
+        {KBinary, VBinary} ->
+            NHeaders = #{
+                <<"content-disposition">> => <<"form-data; name=\"", KBinary/binary, "\"">>,
+                <<"content-type">> => <<"application/x-www-form-urlencoded">>
+            },
+            PartHeaders = cow_multipart:part(Boundary, NHeaders),
+            iolist_to_binary([PartHeaders, VBinary])
+    end.
+
+%% file content type
+file_content_type_headers(Path) ->
+    FileType = filename:extension(Path),
+    case maps:get(FileType, ?FILE_CONTENT_TYPE_MAP, undefined) of
+        undefined ->
```

Comment:
the default Maybe fallback to `octet-stream`

I doubt that cowboy will make its content-type if you don't write content-type?

Reply:
Default json because some api return by `emqx_json:encode(Map)`. Already encoded by self.